chips/sifive/src/plic.rs

```diff
@@ -10,84 +10,187 @@ use kernel::utilities::registers::LocalRegisterCopy;
 use kernel::utilities::registers::{register_bitfields, ReadOnly, ReadWrite};
 use kernel::utilities::StaticRef;
 
+/// Place the register map definition in a private module to disallow direct access to it's
+/// fields from the Plic struct implementation, which should only use a getter/setter with
+/// appropriate bounds set
+
+///    The generic SiFive PLIC specification:
+///    https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc
+///    is defining maximum of 1023 interrupt sources
+
+const MAX_INTERRUPTS: usize = 1023;
+/// maximum number of bit-coded registers, 1 bit per interrupt
+const MAX_BIT_REGS: usize = MAX_INTERRUPTS.div_ceil(32);
+/// PLIC registers for *machine mode* context only at this time.
+/// The spec defines extra sets of registers for additional contexts,
+/// that is supervisor, user and other modes, but these aren't supported
+/// by the current code.
 #[repr(C)]
 pub struct PlicRegisters {
     /// Interrupt Priority Register
     _reserved0: u32,
-    priority: [ReadWrite<u32, priority::Register>; 51],
-    _reserved1: [u8; 3888],
+    priority: [ReadWrite<u32, priority::Register>; MAX_INTERRUPTS],
+    _reserved1: [u8; 0x1000 - (MAX_INTERRUPTS + 1) * 4],
     /// Interrupt Pending Register
-    pending: [ReadOnly<u32>; 2],
-    _reserved2: [u8; 4088],
+    pending: [ReadOnly<u32>; MAX_BIT_REGS],
+    _reserved2: [u8; 0x1000 - MAX_BIT_REGS * 4],
     /// Interrupt Enable Register
-    enable: [ReadWrite<u32>; 2],
-    _reserved3: [u8; 2088952],
+    enable: [ReadWrite<u32>; MAX_BIT_REGS],
+    _reserved3: [u8; 0x20_0000 - 0x2000 - MAX_BIT_REGS * 4],
     /// Priority Threshold Register
     threshold: ReadWrite<u32, priority::Register>,
     /// Claim/Complete Register
     claim: ReadWrite<u32>,
 }
 
+/// Check that the registers are aligned to the PLIC memory map
```

Comment:
n.b. that these kinds of checks should be obviated by @jrvanwhy's eventual Tock registers rewrite, but for now this seems like a good pattern actually!